### PR TITLE
Use local version identifiers to distribute cuda jaxlib wheels.

### PR DESCRIPTION
**Pull request by skye**
This change:

* Updates our jaxlib build scripts to add `+cudaXXX` to the wheel
  version, where XXX is the CUDA version number (e.g. `110`). nocuda
  builds remain unchanged and do not have this extra identifier.

* Adds `generate_release_index.py`, which writes an html page that pip
  can use to find the cuda wheels. (I based this format off of
  wheel PyTorch's index).

* Updates the README to use the new local version identifier + wheel
  index.

The end result is that the command to install cuda wheels is now much
simpler.

I manually made copies of the latest jaxlib 0.1.52 wheels that have
the local version identifiers, so the new installation commands
already work (as well as the old ones, until the next jaxlib release
using the new tooling).

Fow now, I put the html index to the GCP bucket with the wheels. We
can move it to a prettier URL if/when we have one.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -408,14 +408,7 @@ and CUDNN7 installations on your machine (for example, preinstalled on your
...
 ```bash
-# install jaxlib
-PYTHON_VERSION=cp37  # alternatives: cp36, cp37, cp38
-CUDA_VERSION=cuda100  # alternatives: cuda100, cuda101, cuda102, cuda110
-PLATFORM=manylinux2010_x86_64  # alternatives: manylinux2010_x86_64
-BASE_URL='https://storage.googleapis.com/jax-releases'
-pip install --upgrade $BASE_URL/$CUDA_VERSION/jaxlib-0.1.52-$PYTHON_VERSION-none-$PLATFORM.whl
-
-pip install --upgrade jax  # install jax
+pip install --upgrade jax jaxlib==0.1.52+cuda110 -f https://storage.googleapis.com/jax-releases/jax_releases.html
```
Comment (hawkinsp) [inline README.md:411]: You should wait until we do a jaxlib release to make the README change.

File: README.md [hunk 2]
```diff
@@ -444,14 +437,10 @@ Or set the following environment variable before importing JAX:
...
 
-The Python version must match your Python interpreter. There are prebuilt wheels
-for Python 3.6, 3.7, and 3.8; for anything else, you must build from source. Jax
-requires Python 3.6 or above. Jax does not support Python 2 any more.
-
 To try automatic detection of the correct version for your system, you can run:
 
 ```bash
-pip install --upgrade https://storage.googleapis.com/jax-releases/`nvcc -V | sed -En "s/.* release ([0-9]*)\.([0-9]*),.*/cuda\1\2/p"`/jaxlib-0.1.52-`python3 -V | sed -En "s/Python ([0-9]*)\.([0-9]*).*/cp\1\2/p"`-none-manylinux2010_x86_64.whl jax
+pip install --upgrade jax jaxlib==0.1.52+cuda`nvcc -V | sed -En "s/.* release ([0-9]*)\.([0-9]*),.*/\1\2/p"` -f https://storage.googleapis.com/jax-releases/jax_releases.html
```
Comment (hawkinsp) [inline README.md:443]: I think I'd just write this in two lines:

* use `nvcc --version` to find out your CUDA versoin
* then run `...`

Less cryptic and just as easy to understand I think.

File: build/generate_release_index.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generates jax_releases.html package index.
+
+To update public copy, use:
+gsutil cp jax_releases.html gs://jax-releases/
```
Comment (hawkinsp) [inline build/generate_release_index.py:20]: Add this to the playbook.

File: build/generate_release_index.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+HEADER = """
+<!-- Generated by jax/build/generate_release_index.py, do not update manually! -->
+<html>
+<head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"></head>
+<body>
+"""
+
+FOOTER = "</body>\n</html>\n"
+
+ls_output = subprocess.check_output(["gsutil", "ls", "gs://jax-releases/cuda*"])
```
Comment (hawkinsp) [inline build/generate_release_index.py:36]: Maybe check this fails gracefully e.g., if your internets is down or you have no permissions?

**Review by hawkinsp — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (skye) [reply README.md:412]: Why? I manually updated the already-released wheels so this command will still work. I think it's better to do this at all once if we can, so we don't forget to update the README later and break the install instructions.

File: README.md (hunk 2 above)
Comment (skye) [reply README.md]: We basically already have that above. I just removed the auto install, presumably it's now easy enough to find your cuda version and where to put it in the command, like you say.

File: build/generate_release_index.py (hunk 3 above)
Comment (skye) [reply build/generate_release_index.py:20]: I will once this goes in.

File: build/generate_release_index.py (hunk 4 above)
Comment (skye) [reply build/generate_release_index.py:37]: It looks like it prints stderr to the console, so errors from gsutil should be visible. I added some contextual print statements as well.

File: README.md [hunk 5]
```diff
@@ -408,17 +408,11 @@ and CUDNN7 installations on your machine (for example, preinstalled on your
...
-pip install --upgrade jax  # install jax
+pip install --upgrade pip
+pip install --upgrade jax jaxlib==0.1.52+cuda110 -f https://storage.googleapis.com/jax-releases/jax_releases.html
 ```
 
-The library package name must correspond to the version of the existing CUDA
+The jaxlib version must correspond to the version of the existing CUDA
 installation you want to use, with `cuda110` for CUDA 11.0, `cuda102` for CUDA
 10.2, `cuda101` for CUDA 10.1, and `cuda100` for CUDA 10.0. To find your CUDA
 and CUDNN versions, you can run commands like these, depending on your CUDNN
```
Comment (jekbradbury) [inline README.md:418]: This paragraph doesn't say why you need to know your cuDNN version, so I'm not sure it needs to say how to find it?

File: README.md (hunk 1 above)
Comment (hawkinsp) [reply README.md:412]: Ok, that's fine then. I didn't know you had rebuilt the existing release. (Although it still may avoid some versioning confusion if we just make a new release.)

File: README.md (hunk 2 above)
Comment (hawkinsp) [reply README.md]: But my point is I'd just remove this cryptic one-liner. I think it hurts more than it helps.

File: README.md (hunk 5 above)
Comment (skye) [reply README.md:418]: You're right, A+ reading comprehension! I removed the cuDNN bits.

File: README.md (hunk 2 above)
Comment (skye) [reply README.md]: It's already removed :)

File: README.md (hunk 1 above)
Comment (skye) [reply README.md:412]: I didn't rebuild them, I only renamed them (and kept the existing ones as well), but I think that should be enough... it's possible the version reported by `jaxlib.__version__` will include the `+cudaXXX` once we rebuild, which it doesn't now, but I think that's fine. I can cut a new jaxlib release after this changes goes in to verify it all works and avoid any potential issues.

**Pull request merged**